dev/DynamicDependency/DataStore.cpp

```diff
@@ -0,0 +1,167 @@
+    const auto dataSize{ fileSize.QuadPart };
+    THROW_HR_IF(HRESULT_FROM_WIN32(ERROR_INVALID_DATA), dataSize > INT32_MAX);
+    if (dataSize == 0)
+    {
+        // 0-byte file is invalid. Perhaps power was lost when written but before flushed?
+        // 'Fix' it i.e. delete it and report not-found
+        file.reset();
+        std::filesystem::remove(filename);
+        return PackageDependency();
+    }
+
+    const auto bufferSize{ static_cast<DWORD>(dataSize) + 1 };
+    std::unique_ptr<char[]> bufferUtf8{ new char[bufferSize] };
+
+    DWORD bytesRead{};
+    THROW_IF_WIN32_BOOL_FALSE(::ReadFile(file.get(), bufferUtf8.get(), bufferSize, &bytesRead, nullptr));
+    file.reset();
+    bufferUtf8[bytesRead] = '\0';
+    auto json{ bufferUtf8.get() };
+
+    return MddCore::PackageDependency::FromJSON(json);
+}
+
+void MddCore::DataStore::Save(
+    const MddCore::PackageDependency& packageDependency,
+    const MddCreatePackageDependencyOptions options)
+{
+    // Package depednencies with LifetimeKind.Process are never saved to disk. They live in memory and end with the process
+    if (packageDependency.LifetimeKind() == MddPackageDependencyLifetimeKind::Process)
+    {
+        return;
+    }
+
+    auto json{ packageDependency.ToJSONUtf8() };
+
+    auto path{ GetDataStorePath(options) };
+    path /= L"DynamicDependency";
+    std::filesystem::create_directory(path);
+
+    auto filename{ path / (packageDependency.Id() + DataStore::fileExtension) };
+
+    wil::unique_hfile file{ ::CreateFileW(filename.c_str(), GENERIC_WRITE, FILE_SHARE_DELETE, nullptr, CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL, nullptr) };
+    if (!file)
+    {
+        THROW_HR_MSG(HRESULT_FROM_WIN32(GetLastError()), "%ls", filename.c_str());
+    }
+
+    DWORD bytesWritten{};
+    THROW_IF_WIN32_BOOL_FALSE_MSG(::WriteFile(file.get(), json.c_str(), static_cast<DWORD>(json.length()), &bytesWritten, nullptr), "%ls", filename.c_str());
+}
+
+void MddCore::DataStore::Delete(PCWSTR packageDependencyId)
+{
+    std::filesystem::path relativeFilename{ L"DynamicDependency" };
+    relativeFilename /= std::wstring(packageDependencyId) + DataStore::fileExtension;
+
+    auto filename{ GetDataStorePathForUser() };
+    filename /= relativeFilename;
+    if (!DeleteFileIfExists(filename.c_str()))
+    {
+        filename = GetDataStorePathForSystem();
+        filename /= relativeFilename;
+        DeleteFileIfExists(filename.c_str());
+    }
+}
+
+bool MddCore::DataStore::DeleteFileIfExists(PCWSTR filename)
+{
+    if (!::DeleteFileW(filename))
+    {
+        auto const lastError{ GetLastError() };
+        if ((lastError == ERROR_FILE_NOT_FOUND) || (lastError == ERROR_PATH_NOT_FOUND))
+        {
+            return false;
+        }
+        THROW_HR_MSG(HRESULT_FROM_WIN32(lastError), "Error %d deleting file %ls", lastError, filename);
+    }
+    return true;
+}
+
+HANDLE MddCore::DataStore::OpenFileIfExists(
```

Comment:
Can this return the safehandle instead?

Reply:
What's a `safehandle`?

Or do you mean the `wil::unique_hfile`?

Reply:
Right, whatever the appropriate RAII construct is for this handle.

Reply:
Getting some weird compile error. Will consider in next PR